### PR TITLE
Class-based UDFs

### PR DESCRIPTION
**Pull request by sk1p**
Not finished, for discussion of the new interface. Refs #286

### CODE REVIEW

File: src/libertem/io/dataset/base.py [hunk 1]
```diff
@@ -289,7 +289,11 @@ def get_tiles(self, crop_to=None, full_frames=False, mmap=False, dest_dtype="flo
             convert data to this dtype when reading
 
         roi : np.ndarray
-            1d mask that matches the partition shape to limit the region to work on
+            1d mask that matches the dataset navigation shape to limit the region to work on.
```
Comment (uellue) [inline src/libertem/io/dataset/base.py]: Reminder for @sk1p: Update comment

File: src/libertem/udf/base.py [hunk 2]
```diff
@@ -1,6 +1,275 @@
...
+
+    def init_task_data(self, partition):
+        self.task_data = UDFNamespace(self.get_task_data(partition))
+
+    def init_result_buffers(self):
+        self.results = UDFNamespace(self.get_result_buffers())
+
+
+class UDF(UDFBase):
+    def __init__(self, **kwargs):
```
Comment (uellue) [inline src/libertem/udf/base.py]: Draft: This method is called again on the cluster to create copies. For that reason it should not have side effects beyond `self` and only depend on its parameters.

Discussion with @sk1p: Overloading this method with sth fancy may have unwanted side effects.

Documentation: Include a clear example to condition parameters.

File: src/libertem/udf/FEM.py [hunk 3]
```diff
@@ -1,44 +1,42 @@
...
+        kwargs = {
+            'mask': mask,
+        }
+        return kwargs
+
+    def process_frame(self, frame):
+        self.results.intensity[:] = np.std(frame[self.task_data.mask == 1])
+
+
+def run_fem(ctx, dataset, center, rad_in, rad_out, roi=None):
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/udf/FEM.py]: Hmm, I would say, we should defer this change until we have the integration with the web API, which should also handle multi-stage UDFs for scripting. That means UDFs developed for 0.2 will need a little work to be available in the web API, but I think that's fine.

Maybe this becomes more clear if we port over the blobfinder UDF to the new interface.

**Comment by sk1p** [comment]
I addressed most of the comments, only the async/web API stuff remains. Also, the `full_frames=False` tiled processing is not finished, see #310 for a remark about possible interface changes there.

Do you want to port over the blobfinder UDFs, or should I do this?

File: src/libertem/udf/base.py [hunk 4]
```diff
@@ -9,54 +278,64 @@ def check_cast(fromvar, tovar):
...
+        return self._udf.results, partition
+
+    def run_for_dataset(self, dataset, executor, roi):
+        self._udf.init_result_buffers()
+        self._udf.allocate_for_full(dataset, roi)
+
+        tasks = self._make_udf_tasks(dataset, roi)
+        cancel_id = str(uuid.uuid4())
+
+        for part_results, partition in executor.run_tasks(tasks, cancel_id):
```
[the remark this answers is not in the document]
Comment (sk1p) [reply src/libertem/udf/base.py]: There is now a first prototype of an async version of this function available, including integration with web API

**Comment by sk1p** [comment]
`udf_v2_wip` is now merged!

**Pull request closed without merging**